{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Code used to rearrange output files for ZikV variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: plyr\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.0 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.2     \u001b[32m✔\u001b[39m \u001b[34mdplyr  \u001b[39m 1.0.2\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.0.4     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 1.4.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32marrange()\u001b[39m   masks \u001b[34mplyr\u001b[39m::arrange()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompact()\u001b[39m   masks \u001b[34mplyr\u001b[39m::compact()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mcount()\u001b[39m     masks \u001b[34mplyr\u001b[39m::count()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfailwith()\u001b[39m  masks \u001b[34mplyr\u001b[39m::failwith()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m    masks \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mid()\u001b[39m        masks \u001b[34mplyr\u001b[39m::id()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m       masks \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mmutate()\u001b[39m    masks \u001b[34mplyr\u001b[39m::mutate()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mrename()\u001b[39m    masks \u001b[34mplyr\u001b[39m::rename()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarise()\u001b[39m masks \u001b[34mplyr\u001b[39m::summarise()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32msummarize()\u001b[39m masks \u001b[34mplyr\u001b[39m::summarize()\n",
      "\n",
      "\n",
      "Attaching package: ‘reshape2’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:tidyr’:\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘glue’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# import used packages\n",
    "require('plyr')\n",
    "library('tidyr')\n",
    "library('tidyverse')\n",
    "library('reshape2')\n",
    "library('glue')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inputs to run script\n",
    "wkdir = \"/home/kate/Lab/zika_files/ReAlignment/SubVariants/\" # set working directory\n",
    "\n",
    "input_vars = c(\"ZIKA_1\",\"ZIKA_0.75\",\"ZIKA_0.50\",\"ZIKA_0.25\",\"ZIKA_0.10\",\"ZIKA_0.05\")\n",
    "\n",
    "ref = 'MR766'\n",
    "\n",
    "minfreqs = c(0.01, 0.02, 0.03, 0.05, 0.1)\n",
    "\n",
    "coverage = 200\n",
    "\n",
    "metafile = '/home/kate/Lab/zika_files/ReAlignment/MetadataZika/Zika_Metadata_v4.csv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(wkdir)\n",
    "meta = read.csv(file=metafile,header=T,sep=\",\",na.strings = c('nan'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate new directory if not present to store updated variant files in\n",
    "if (!dir.exists(glue(\"{wkdir}SubVar_Rearranged\"))) {\n",
    "      dir.create(glue(\"{wkdir}SubVar_Rearranged\"))\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generating a dataframe with gene information from ZikV website\n",
    "Gene_ID = c('C','pr','M','E','NS1','NS2A','NS2B','NS3','NS4A','2K','NS4B','NS5')\n",
    "\n",
    "Gene_start = c(107,473,752,977,2477,3533,4211,4601,6452,6833,6902,7655)\n",
    "\n",
    "Gene_End = c(472,751,976,2476,3532,4210,4600,6451,6832,6901,7654,10363) #ns3 ns4a off by 100 missing\n",
    "\n",
    "Gene_df = cbind(Gene_ID,Gene_start,Gene_End)\n",
    "\n",
    "Gene_df = as.data.frame(Gene_df)\n",
    "\n",
    "Gene_df$Gene_start = (as.numeric(as.character(Gene_df$Gene_start)))-106\n",
    "\n",
    "Gene_df$Gene_End = as.numeric(as.character(Gene_df$Gene_End))-106\n",
    "\n",
    "Genes = data.frame(gene = as.character(), position = as.numeric())\n",
    "\n",
    "for (id in Gene_ID){\n",
    "  gene_Range = filter(Gene_df, Gene_ID == id)\n",
    "  gene_Range=droplevels(gene_Range)\n",
    "  gene_ids = c(rep(id, (gene_Range$Gene_End + 1 - gene_Range$Gene_start)))\n",
    "  ntpos = c(seq(gene_Range$Gene_start, gene_Range$Gene_End, by =1))\n",
    "  genes_df = cbind(gene_ids,ntpos)\n",
    "  genes_df = as.data.frame(genes_df)\n",
    "  Genes = rbind(Genes, genes_df)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "Rearrange = function(filename, subs, savedir, meta_df, genedf, coverage, frq){\n",
    "    \n",
    "    print(filename)\n",
    "    \n",
    "    # read in the file as a dataframe\n",
    "    mydata=read.csv(file=filename,header=T,sep=\",\",na.strings = c('nan'))\n",
    "\n",
    "    print(dim(mydata))\n",
    "    \n",
    "    # merge with the metadata to have all info in one place\n",
    "    mydata_meta = merge(mydata, meta_df,by.x='sample',by.y='name')\n",
    "    \n",
    "    print(dim(mydata_meta))\n",
    "    \n",
    "    # add gene name information here by merging using nucleotide position\n",
    "    main_df = merge(mydata_meta, genedf, by='ntpos')\n",
    "\n",
    "    print(dim(main_df))\n",
    "    \n",
    "    # filter out minor variant information\n",
    "    min_df = filter(main_df, majmin=='minor') %>%\n",
    "          select(ntpos,sample,segment,nt,majmin,\n",
    "                 freq,aa,codon,nonsyn,binocheck,\n",
    "                 totalcount,mouse_id,transmission,\n",
    "                 tissue,sex,type,id2,id,gene_ids,\n",
    "                 experiment, full,pcr1) %>% \n",
    "    droplevels()\n",
    "    \n",
    "    # rename columns \n",
    "    colnames(min_df)= c('ntpos','sample','segment',\n",
    "                        'minornt','majmin','minorfreq',\n",
    "                        'minoraa','minorcodon','nonsyn',\n",
    "                        'binocheck','totalcount','mouse_id',\n",
    "                        'transmission','tissue','sex',\n",
    "                        'type','id2','id','gene_ids',\n",
    "                        'experiment','full','pcr1')\n",
    "    \n",
    "    \n",
    "    # filter out major information for variant positions: \n",
    "    maj_df = filter(main_df, majmin =='major')  %>%\n",
    "      select(sample, segment, ntpos, nt, aa, codon,freq, aapos) %>% \n",
    "    droplevels()\n",
    "\n",
    "    # change column names\n",
    "    colnames(maj_df) = c('sample','segment','ntpos','majornt','majoraa',\n",
    "                     'majorcodon','majorfreq','aapos')\n",
    "\n",
    "    # merge the minor and major information using sample and ntpos\n",
    "    df_merge = merge(min_df, maj_df, by=c('sample','ntpos','segment')) %>%\n",
    "        select(sample,ntpos,majornt,majorfreq,minornt,minorfreq,\n",
    "             aapos,majorcodon,minorcodon,majoraa,minoraa,nonsyn,\n",
    "               gene_ids,mouse_id,transmission,tissue,sex,type,\n",
    "               id2,id,binocheck,segment,experiment, full, pcr1) %>%\n",
    "      droplevels()\n",
    "\n",
    "    print(dim(df_merge))\n",
    "    \n",
    "    # write the merged file to a new csv file\n",
    "    # save in generated sub directory\n",
    "    write.csv(df_merge, \n",
    "          file = glue(\"{savedir}/ZikaVariants.{subs}.{coverage}.{frq}.csv\"),\n",
    "              row.names=F)\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"ZIKA_1\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_1.all.listVariants.0.01.csv\n",
      "[1] 3470   12\n",
      "[1] 3470   26\n",
      "[1] 3470   27\n",
      "[1] 1735   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_1.all.listVariants.0.02.csv\n",
      "[1] 1820   12\n",
      "[1] 1820   26\n",
      "[1] 1820   27\n",
      "[1] 910  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_1.all.listVariants.0.03.csv\n",
      "[1] 1086   12\n",
      "[1] 1086   26\n",
      "[1] 1086   27\n",
      "[1] 543  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_1.all.listVariants.0.05.csv\n",
      "[1] 500  12\n",
      "[1] 500  26\n",
      "[1] 500  27\n",
      "[1] 250  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_1.all.listVariants.0.1.csv\n",
      "[1] 136  12\n",
      "[1] 136  26\n",
      "[1] 136  27\n",
      "[1] 68 25\n",
      "[1] \"ZIKA_0.75\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_0.75.all.listVariants.0.01.csv\n",
      "[1] 3398   12\n",
      "[1] 3398   26\n",
      "[1] 3398   27\n",
      "[1] 1699   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_0.75.all.listVariants.0.02.csv\n",
      "[1] 1824   12\n",
      "[1] 1824   26\n",
      "[1] 1824   27\n",
      "[1] 912  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_0.75.all.listVariants.0.03.csv\n",
      "[1] 1096   12\n",
      "[1] 1096   26\n",
      "[1] 1096   27\n",
      "[1] 548  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_0.75.all.listVariants.0.05.csv\n",
      "[1] 502  12\n",
      "[1] 502  26\n",
      "[1] 502  27\n",
      "[1] 251  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_0.75.all.listVariants.0.1.csv\n",
      "[1] 140  12\n",
      "[1] 140  26\n",
      "[1] 140  27\n",
      "[1] 70 25\n",
      "[1] \"ZIKA_0.50\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_0.50.all.listVariants.0.01.csv\n",
      "[1] 3370   12\n",
      "[1] 3370   26\n",
      "[1] 3370   27\n",
      "[1] 1685   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_0.50.all.listVariants.0.02.csv\n",
      "[1] 1832   12\n",
      "[1] 1832   26\n",
      "[1] 1832   27\n",
      "[1] 916  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_0.50.all.listVariants.0.03.csv\n",
      "[1] 1080   12\n",
      "[1] 1080   26\n",
      "[1] 1080   27\n",
      "[1] 540  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_0.50.all.listVariants.0.05.csv\n",
      "[1] 510  12\n",
      "[1] 510  26\n",
      "[1] 510  27\n",
      "[1] 255  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_0.50.all.listVariants.0.1.csv\n",
      "[1] 138  12\n",
      "[1] 138  26\n",
      "[1] 138  27\n",
      "[1] 69 25\n",
      "[1] \"ZIKA_0.25\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_0.25.all.listVariants.0.01.csv\n",
      "[1] 3216   12\n",
      "[1] 3216   26\n",
      "[1] 3216   27\n",
      "[1] 1608   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_0.25.all.listVariants.0.02.csv\n",
      "[1] 1810   12\n",
      "[1] 1810   26\n",
      "[1] 1810   27\n",
      "[1] 905  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_0.25.all.listVariants.0.03.csv\n",
      "[1] 1062   12\n",
      "[1] 1062   26\n",
      "[1] 1062   27\n",
      "[1] 531  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_0.25.all.listVariants.0.05.csv\n",
      "[1] 492  12\n",
      "[1] 492  26\n",
      "[1] 492  27\n",
      "[1] 246  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_0.25.all.listVariants.0.1.csv\n",
      "[1] 138  12\n",
      "[1] 138  26\n",
      "[1] 138  27\n",
      "[1] 69 25\n",
      "[1] \"ZIKA_0.10\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_0.10.all.listVariants.0.01.csv\n",
      "[1] 2886   12\n",
      "[1] 2886   26\n",
      "[1] 2886   27\n",
      "[1] 1443   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_0.10.all.listVariants.0.02.csv\n",
      "[1] 1776   12\n",
      "[1] 1776   26\n",
      "[1] 1776   27\n",
      "[1] 888  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_0.10.all.listVariants.0.03.csv\n",
      "[1] 1086   12\n",
      "[1] 1086   26\n",
      "[1] 1086   27\n",
      "[1] 543  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_0.10.all.listVariants.0.05.csv\n",
      "[1] 512  12\n",
      "[1] 512  26\n",
      "[1] 512  27\n",
      "[1] 256  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_0.10.all.listVariants.0.1.csv\n",
      "[1] 144  12\n",
      "[1] 144  26\n",
      "[1] 144  27\n",
      "[1] 72 25\n",
      "[1] \"ZIKA_0.05\"\n",
      "[1] 0.01\n",
      "sub_linegraphs/ZIKA_0.05.all.listVariants.0.01.csv\n",
      "[1] 2766   12\n",
      "[1] 2766   26\n",
      "[1] 2766   27\n",
      "[1] 1383   25\n",
      "[1] 0.02\n",
      "sub_linegraphs/ZIKA_0.05.all.listVariants.0.02.csv\n",
      "[1] 1684   12\n",
      "[1] 1684   26\n",
      "[1] 1684   27\n",
      "[1] 842  25\n",
      "[1] 0.03\n",
      "sub_linegraphs/ZIKA_0.05.all.listVariants.0.03.csv\n",
      "[1] 1056   12\n",
      "[1] 1056   26\n",
      "[1] 1056   27\n",
      "[1] 528  25\n",
      "[1] 0.05\n",
      "sub_linegraphs/ZIKA_0.05.all.listVariants.0.05.csv\n",
      "[1] 456  12\n",
      "[1] 456  26\n",
      "[1] 456  27\n",
      "[1] 228  25\n",
      "[1] 0.1\n",
      "sub_linegraphs/ZIKA_0.05.all.listVariants.0.1.csv\n",
      "[1] 146  12\n",
      "[1] 146  26\n",
      "[1] 146  27\n",
      "[1] 73 25\n"
     ]
    }
   ],
   "source": [
    "# iterate through the various subsampled variant files at the diff. frequencies\n",
    "\n",
    "for (vartype in input_vars){\n",
    "    print(vartype)\n",
    "    \n",
    "    for (freq in minfreqs){\n",
    "        print(freq)\n",
    "    \n",
    "        # import file name for each subset of variants\n",
    "        filename = glue(\"sub_linegraphs/{vartype}.all.listVariants.{freq}.csv\")\n",
    "    \n",
    "        \n",
    "        # generate new csv files for these\n",
    "        Rearrange(filename, vartype, glue(\"{wkdir}SubVar_Rearranged\"), \n",
    "                  meta, Genes, coverage, freq)    \n",
    "  }  \n",
    "    \n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
